test/cassandra/container.spec.ts

```diff
@@ -0,0 +1,93 @@
+import "expect-puppeteer";
+import crypto from 'crypto'
+
+jest.setTimeout(300000);
+
+describe('Collection Add and Delete Cassandra spec', () => {
+  it('creates a collection', async () => {
+    try {
+      const keyspaceId = `keyspaceid${crypto.randomBytes(8).toString("hex")}`;
+      const tableId = `tableid${crypto.randomBytes(3).toString('hex')}`;
+      const prodUrl = "https://localhost:1234/hostedExplorer.html";
+      page.goto(prodUrl);
+
+      // log in with connection string
+      const handle = await page.waitForSelector('iframe');
+      const frame = await handle.contentFrame();
+      await frame.waitFor('div > p.switchConnectTypeText', { visible: true });
+      await frame.click('div > p.switchConnectTypeText');
+      const connStr = process.env.CASSANDRA_CONNECTION_STRING;
+      await frame.type("input[class='inputToken']", connStr);
+      await frame.click("input[value='Connect']");
+
+      // create new table
+      await frame.waitFor('button[data-test="New Table"]', { visible: true });
+      await frame.waitForSelector('div[class="splashScreen"] > div[class="title"]', { visible: true });
+      await frame.click('button[data-test="New Table"]');      
+
+      // type keyspace id
+      await frame.waitFor('input[id="keyspace-id"]', { visible: true });
+      await frame.type('input[id="keyspace-id"]', keyspaceId);      
+
+      // type table id
+      await frame.waitFor('input[class="textfontclr"]');
+      await frame.type('input[class="textfontclr"]', tableId); 
+
+      // click submit
+      await frame.waitFor('#cassandraaddcollectionpane > div > form > div.paneFooter > div > input');
+      await frame.click('#cassandraaddcollectionpane > div > form > div.paneFooter > div > input');
+
+      // open database menu
+      await frame.waitFor(`div[data-test="${keyspaceId}"]`);
+      await frame.waitForSelector('div[class="splashScreen"] > div[class="title"]', { visible: true });
+
+      await frame.click(`div[data-test="${keyspaceId}"]`);
+      await frame.waitFor(`span[title="${tableId}"]`);
+      
+      // delete container
+
+      // click context menu for container
+      await frame.waitFor(`div[data-test="${tableId}"] > div > button`);
+      await frame.click(`div[data-test="${tableId}"] > div > button`);
+
+      // click delete container
+      await frame.waitForSelector('body > div.ms-Layer.ms-Layer--fixed');
+      const elements = await frame.$$('span[class="treeComponentMenuItemLabel"]')
+      await elements[0].click()
+
+      // confirm delete container
+      await frame.type('input[data-test="confirmCollectionId"]', tableId.trim());
+
+      // click delete
+      await frame.click('input[data-test="deleteCollection"]');
+      await frame.waitFor(5000);
```

Comment:
Let's define all these as constants. Reuse the same constant if possible for the all the `waitFor`. We make it one dial in case we want to fine tune and speed up the test (by decreasing the constant).
Here I noticed that we wait for selector, do we actually need the `waitFor` or can we reduce to 1s (presumably for the ui to refresh to some state?). I'm just concerned that once we have many of these tests it will take for ever...

Reply:
These `waitFor` calls end up making the tests much less brittle but can be reduced. The tests try to click elements before they seem to show up without them. In this particular case, when we click `Delete` to confirm deleting a container, there is significant lag for the UI to show the loading state (and we are waiting for the "ready" state). So the "ready" state is already present, and the test moves on before we ever hit the loading state. I can reduce this I think.

As for speed, I'll get these running in parallel when I'm working on the last one to hopefully avoid any issues with CI taking forever.

Reply:
The best solution is probably to wait for the loading state, then the ready state!